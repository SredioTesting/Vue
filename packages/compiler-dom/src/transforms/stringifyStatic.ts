/**
 * This module is Node-only.
 */
import {
  CREATE_STATIC,
  createCallExpression,
  ElementNode,
  ElementTypes,
  ExpressionNode,
  HoistTransform,
  JSChildNode,
  NodeTypes,
  PlainElementNode,
  SimpleExpressionNode,
  TemplateChildNode,
  TextCallNode,
  TransformContext
} from '@vue/compiler-core'
import {
  escapeHtml,
  isKnownAttr,
  isString,
  isSymbol,
  isVoidTag,
  normalizeClass,
  normalizeStyle,
  stringifyStyle,
  toDisplayString
} from '@vue/shared'

export const enum StringifyThresholds {
  ELEMENT_WITH_BINDING_COUNT = 5,
  NODE_COUNT = 20
}

type StringifiableNode = PlainElementNode | TextCallNode

/**
 * Turn eligible hoisted static trees into stringified static nodes, e.g.
 *
 * ```js
 * const _hoisted_1 = createStaticVNode(`<div class="foo">bar</div>`)
 * ```
 *
 * A single static vnode can contain stringified content for **multiple**
 * consecutive nodes (element and plain text), called a "chunk".
 * `@vue/runtime-dom` will create the content via innerHTML in a hidden
 * container element and insert all the nodes in place. The call must also
 * provide the number of nodes contained in the chunk so that during hydration
 * we can know how many nodes the static vnode should adopt.
 *
 * The optimization scans a children list that contains hoisted nodes, and
 * tries to find the largest chunk of consecutive hoisted nodes before running
 * into a non-hoisted node or the end of the list. A chunk is then converted
 * into a single static vnode and replaces the hoisted expression of the first
 * node in the chunk. Other nodes in the chunk are considered "merged" and
 * therefore removed from both the hoist list and the children array.
 *
 * This optimization is only performed in Node.js.
 */
export const stringifyStatic: HoistTransform = (children, context, parent) => {
  let nc = 0 // current node count
  let ec = 0 // current element with binding count
  const currentChunk: StringifiableNode[] = []
  const chunks: (JSChildNode | TemplateChildNode)[] = []
  const parentIsHoistedNode = getHoistedNode(parent as TemplateChildNode)

  const pushChunk = () => {
    if (
      nc >= StringifyThresholds.NODE_COUNT ||
      ec >= StringifyThresholds.ELEMENT_WITH_BINDING_COUNT
    ) {
      // combine all currently eligible nodes into a single static vnode call
      const staticCall = createCallExpression(context.helper(CREATE_STATIC), [
        JSON.stringify(
          currentChunk.map(node => stringifyNode(node, context)).join('')
        ),
        // the 2nd argument indicates the number of DOM nodes this static vnode
        // will insert / hydrate
        String(currentChunk.length)
      ])

      if (parentIsHoistedNode) {
        // if parent is hoisted, just append staticCall instead of hoisted vnode list
        chunks.push(staticCall)
      } else {
        // if parent isn't hoisted, this mean is hoisted node self.
        // so need remove chunk nodes, re-create hoist for staticCall
        for (let i = 0; i < currentChunk.length; i++) {
          removeHoist(currentChunk[i], context)
        }
        chunks.push(context.hoist(staticCall))
      }
      currentChunk.length = 0
    }
    // if the nodes of currentChunk can't static, should append to chunks
    if (currentChunk.length) {
      chunks.push(...currentChunk)
    }
  }

  let i = 0
  for (; i < children.length; i++) {
    const child = children[i]
    if (getHoistedNode(child) || parentIsHoistedNode) {
      // presence of hoisted means child must be a stringifiable node
      const result = analyzeNode(child)
      if (result) {
        // node is stringifiable, record state
        nc += result[0]
        ec += result[1]
        currentChunk.push(child as StringifiableNode)
        continue
      }
    }
    // we only reach here if we ran into a node that is not stringifiable
    // check if currently analyzed nodes meet criteria for stringification.
    pushChunk()
    // current node should append to chunks
    chunks.push(child)
    // reset state
    nc = 0
    ec = 0
  }
  // maybe current chunk has children
  pushChunk()
  if (parentIsHoistedNode) {
    (parent as any).codegenNode.hoisted!.children = chunks as TemplateChildNode[]
  } else {
    parent.children = chunks as TemplateChildNode[]
  }
}

const getHoistedNode = (node: TemplateChildNode) =>
  ((node.type === NodeTypes.ELEMENT && node.tagType === ElementTypes.ELEMENT) ||
    node.type == NodeTypes.TEXT_CALL) &&
  node.codegenNode &&
  node.codegenNode.type === NodeTypes.SIMPLE_EXPRESSION &&
  node.codegenNode.hoisted

const dataAriaRE = /^(data|aria)-/
const isStringifiableAttr = (name: string) => {
  return isKnownAttr(name) || dataAriaRE.test(name)
}

const removeHoist = (node: StringifiableNode, context: TransformContext) => {
  const hoistToRemove = (node.codegenNode as SimpleExpressionNode).hoisted!
  context.hoists.splice(context.hoists.indexOf(hoistToRemove), 1)
}

/**
 * for a hoisted node, analyze it and return:
 * - false: bailed (contains runtime constant)
 * - [nc, ec] where
 *   - nc is the number of nodes inside
 *   - ec is the number of element with bindings inside
 */
function analyzeNode(node: TemplateChildNode): [number, number] | false {
  if (node.type === NodeTypes.TEXT_CALL) {
    return [1, 0]
  }
  if (node.type !== NodeTypes.ELEMENT) {
    return false
  }

  let nc = 1 // node count
  let ec = node.props.length > 0 ? 1 : 0 // element w/ binding count
  let bailed = false
  const bail = (): false => {
    bailed = true
    return false
  }

  // TODO: check for cases where using innerHTML will result in different
  // output compared to imperative node insertions.
  // probably only need to check for most common case
  // i.e. non-phrasing-content tags inside `<p>`
  function walk(node: ElementNode): boolean {
    for (let i = 0; i < node.props.length; i++) {
      const p = node.props[i]
      // bail on non-attr bindings
      if (p.type === NodeTypes.ATTRIBUTE && !isStringifiableAttr(p.name)) {
        return bail()
      }
      if (p.type === NodeTypes.DIRECTIVE && p.name === 'bind') {
        // bail on non-attr bindings
        if (
          p.arg &&
          (p.arg.type === NodeTypes.COMPOUND_EXPRESSION ||
            (p.arg.isStatic && !isStringifiableAttr(p.arg.content)))
        ) {
          return bail()
        }
      }
    }
    for (let i = 0; i < node.children.length; i++) {
      nc++
      if (nc >= StringifyThresholds.NODE_COUNT) {
        return true
      }
      const child = node.children[i]
      if (child.type === NodeTypes.ELEMENT) {
        if (child.props.length > 0) {
          ec++
          if (ec >= StringifyThresholds.ELEMENT_WITH_BINDING_COUNT) {
            return true
          }
        }
        walk(child)
        if (bailed) {
          return false
        }
      }
    }
    return true
  }

  return walk(node) ? [nc, ec] : false
}

function stringifyNode(
  node: string | TemplateChildNode,
  context: TransformContext
): string {
  if (isString(node)) {
    return node
  }
  if (isSymbol(node)) {
    return ``
  }
  switch (node.type) {
    case NodeTypes.ELEMENT:
      return stringifyElement(node, context)
    case NodeTypes.TEXT:
      return escapeHtml(node.content)
    case NodeTypes.COMMENT:
      return `<!--${escapeHtml(node.content)}-->`
    case NodeTypes.INTERPOLATION:
      return escapeHtml(toDisplayString(evaluateConstant(node.content)))
    case NodeTypes.COMPOUND_EXPRESSION:
      return escapeHtml(evaluateConstant(node))
    case NodeTypes.TEXT_CALL:
      return stringifyNode(node.content, context)
    default:
      // static trees will not contain if/for nodes
      return ''
  }
}

function stringifyElement(
  node: ElementNode,
  context: TransformContext
): string {
  let res = `<${node.tag}`
  for (let i = 0; i < node.props.length; i++) {
    const p = node.props[i]
    if (p.type === NodeTypes.ATTRIBUTE) {
      res += ` ${p.name}`
      if (p.value) {
        res += `="${escapeHtml(p.value.content)}"`
      }
    } else if (p.type === NodeTypes.DIRECTIVE && p.name === 'bind') {
      // constant v-bind, e.g. :foo="1"
      let evaluated = evaluateConstant(p.exp as SimpleExpressionNode)
      const arg = p.arg && (p.arg as SimpleExpressionNode).content
      if (arg === 'class') {
        evaluated = normalizeClass(evaluated)
      } else if (arg === 'style') {
        evaluated = stringifyStyle(normalizeStyle(evaluated))
      }
      res += ` ${(p.arg as SimpleExpressionNode).content}="${escapeHtml(
        evaluated
      )}"`
    }
  }
  if (context.scopeId) {
    res += ` ${context.scopeId}`
  }
  res += `>`
  for (let i = 0; i < node.children.length; i++) {
    res += stringifyNode(node.children[i], context)
  }
  if (!isVoidTag(node.tag)) {
    res += `</${node.tag}>`
  }
  return res
}

// __UNSAFE__
// Reason: eval.
// It's technically safe to eval because only constant expressions are possible
// here, e.g. `{{ 1 }}` or `{{ 'foo' }}`
// in addition, constant exps bail on presence of parens so you can't even
// run JSFuck in here. But we mark it unsafe for security review purposes.
// (see compiler-core/src/transformExpressions)
function evaluateConstant(exp: ExpressionNode): string {
  if (exp.type === NodeTypes.SIMPLE_EXPRESSION) {
    return new Function(`return ${exp.content}`)()
  } else {
    // compound
    let res = ``
    exp.children.forEach(c => {
      if (isString(c) || isSymbol(c)) {
        return
      }
      if (c.type === NodeTypes.TEXT) {
        res += c.content
      } else if (c.type === NodeTypes.INTERPOLATION) {
        res += toDisplayString(evaluateConstant(c.content))
      } else {
        res += evaluateConstant(c)
      }
    })
    return res
  }
}
