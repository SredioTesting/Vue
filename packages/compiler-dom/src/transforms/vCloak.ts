import { NOOP } from '@vue/shared'
import { createStructuralDirectiveTransform } from '@vue/compiler-core'

export const transformCloak = createStructuralDirectiveTransform('cloak', NOOP)
