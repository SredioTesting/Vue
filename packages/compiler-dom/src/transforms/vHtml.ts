import {
  createCompilerError,
  ErrorCodes,
  createSimpleExpression,
  NodeTypes,
  NodeTransform,
  DirectiveNode
} from '@vue/compiler-core'

export const transformHtml: NodeTransform = (node, context) => {
  if (node.type === NodeTypes.ELEMENT) {
    const prop = node.props.find(
      x => x.type === NodeTypes.DIRECTIVE && x.name === 'html'
    ) as DirectiveNode | undefined
    if (prop) {
      prop.name = `bind`
      prop.arg = createSimpleExpression(`innerHTML`, true, prop.loc)

      if (!prop.exp || !prop.exp.loc.source.trim()) {
        context.onError(
          createCompilerError(ErrorCodes.X_V_HTML_NO_EXPRESSION, prop.loc)
        )
      }
      if (node.children.length > 0) {
        // remove all the children, since they will be overridden by the `innerHTML`
        node.children = []
        if (__DEV__) {
          console.warn(`"v-html" replaced children on "${node.tag}" element`)
        }
      }
    }
  }
}
