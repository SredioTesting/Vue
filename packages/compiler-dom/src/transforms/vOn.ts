import {
  transformOn as baseTransform,
  DirectiveTransform,
  createObjectProperty,
  createCallExpression
} from '@vue/compiler-core'
import { V_ON_MODIFIERS_GUARD } from '../runtimeHelpers'

export const transformOn: DirectiveTransform = (dir, node, context) => {
  const { modifiers } = dir
  let baseResult = baseTransform(dir, node, context)
  if (!modifiers.length) return baseResult
  let exp = baseResult.props[0].value
  return {
    props: [
      createObjectProperty(
        baseResult.props[0].key,
        createCallExpression(context.helper(V_ON_MODIFIERS_GUARD), [
          exp,
          JSON.stringify(dir.modifiers)
        ])
      )
    ],
    needRuntime: false
  }
}
