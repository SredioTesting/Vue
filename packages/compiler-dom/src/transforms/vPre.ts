import { NodeTypes, NodeTransform, DirectiveNode, SimpleExpressionNode } from '@vue/compiler-core'

export const transformPre: NodeTransform = (node, context) => {
  if (node.type === NodeTypes.ELEMENT) {
    const index = node.props.findIndex(
      x => x.type === NodeTypes.DIRECTIVE && x.name === 'pre'
    );
    if (index >= 0) {
      const prop = node.props[index] as DirectiveNode
      if(__DEV__ && prop.exp && !!(prop.exp as SimpleExpressionNode).content.trim()){
        console.warn(`Unexpected expression on "v-pre".`, prop.exp)
      }
      
      node.props.splice(index, 1);

      const start = node.children[0].loc.start
      const end = node.children[node.children.length - 1].loc.end

      const loc = {
        start: start,
        end: end,
        source: context.root.loc.source.slice(
          start.offset,
          end.offset
        )
      }

      // override the children with only the source of the node content
      node.children = [
        {
          type: NodeTypes.TEXT,
          content: loc.source,
          loc: loc,
          isEmpty: false
        }
      ]
    }
  }
}
