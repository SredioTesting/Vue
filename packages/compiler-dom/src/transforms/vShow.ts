import {
  createCompilerError,
  ErrorCodes,
  DirectiveTransform,
  createSimpleExpression,
  createCompoundExpression,
  createObjectExpression,
  createObjectProperty,
  SimpleExpressionNode
} from '@vue/compiler-core'

export const transformShow: DirectiveTransform = (dir, context) => {
  const { exp, loc } = dir
  if (!exp || (exp && !(exp as SimpleExpressionNode).content.trim())) {
    context.onError(createCompilerError(ErrorCodes.X_V_SHOW_NO_EXPRESSION, loc))
  }

  return {
    props: createObjectProperty(
      'style',
      createCompoundExpression([
        exp!,
        '?',
        createObjectExpression([
          createObjectProperty('display', createSimpleExpression('none', true))
        ]),
        ':',
        createObjectExpression([])
      ])
    ),
    needRuntime: false
  }
}
