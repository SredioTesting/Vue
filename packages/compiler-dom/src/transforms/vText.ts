// TODO
import {
  createStructuralDirectiveTransform,
  createCompilerError,
  ErrorCodes,
  createSimpleExpression,
  SimpleExpressionNode,
  NodeTypes
} from '@vue/compiler-core'

export const transformText = createStructuralDirectiveTransform(
  'text',
  (node, dir, context) => {
    if (!dir.exp || !(dir.exp as SimpleExpressionNode).content.trim()) {
      const loc = dir.exp ? dir.exp.loc : node.loc
      context.onError(
        createCompilerError(ErrorCodes.X_V_TEXT_NO_EXPRESSION, dir.loc)
      )
      dir.exp = createSimpleExpression('', true, loc)
    }

    // v-show can't be used outside of an element
    if (node.type !== NodeTypes.ELEMENT) {
      createCompilerError(ErrorCodes.X_V_TEXT_UNEXPECTED_USAGE, dir.loc)
    }

    // add prop textContent
    node.props.push({
      type: NodeTypes.DIRECTIVE,
      name: `bind`,
      arg: createSimpleExpression(`textContent`, true, dir.loc),
      exp: dir.exp,
      modifiers: [],
      loc: dir.loc
    })
  }
)
