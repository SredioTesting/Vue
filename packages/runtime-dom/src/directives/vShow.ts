import { Directive } from '@vue/runtime-core'

type ShowGuardElement = ElementCSSInlineStyle & {
  __vOriginalDisplay: string | null
}

//TODO: interaction with transition module
export const vShowGuard: Directive<ShowGuardElement> = {
  beforeMount: (el, { value }) => {
    setOriginalDisplay(el)
    setElementDisplayByValue(el, value)
  },
  updated: (el, { value, oldValue }) => {
    if (!value === !oldValue) return
    setElementDisplayByValue(el, value)
  },
  beforeUnmount: el => {
    setElementDisplayByValue(el, true)
  }
}

function setOriginalDisplay(el: ShowGuardElement): void {
  el.__vOriginalDisplay = el.style.display === 'none' ? '' : el.style.display
}

function setElementDisplayByValue(el: ShowGuardElement, value: any): void {
  el.style.display = value ? el.__vOriginalDisplay : 'none'
}
