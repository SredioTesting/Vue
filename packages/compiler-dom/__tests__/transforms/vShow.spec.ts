import {
  parse,
  transform,
  CompilerOptions,
  ElementNode,
  NodeTypes,
  createObjectExpression,
  createObjectProperty,
  createSimpleExpression
} from '@vue/compiler-core'
import { transformShow } from '../../src/transforms/vShow'
import { transformStyle } from '../../src/transforms/transformStyle'

function parseWithShowTransform(
  template: string,
  options: CompilerOptions = {}
) {
  const ast = parse(template, options)
  transform(ast, {
    nodeTransforms: [
      transformStyle,
      // NOTE: transformShow must come come after the style
      transformShow
    ],
    ...options
  })
  return {
    root: ast,
    node: ast.children[0] as ElementNode
  }
}

describe('compiler: `v-show` transform', () => {
  it('should add style directive', () => {
    const { node } = parseWithShowTransform(`<div v-show="true"/>`)

    expect(node.props[0]).toMatchObject({
      type: NodeTypes.DIRECTIVE,
      name: `bind`,
      arg: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: `style`,
        isStatic: true
      },
      exp: {
        type: NodeTypes.COMPOUND_EXPRESSION,
        children: [
          {
            type: NodeTypes.SIMPLE_EXPRESSION,
            isStatic: false,
            content: 'true'
          },
          "?",
          createObjectExpression([
            createObjectProperty('display', createSimpleExpression('none', true))
          ]),
          ":",
          createObjectExpression([])
        ]
      }
    })
  })

  describe('style', () => {
    it('should append to style', () => {
      const { node } = parseWithShowTransform(
        `<div style="display:flex" v-show="true"/>`
      )

      expect(node.props).toMatchObject([
        {
          type: NodeTypes.DIRECTIVE,
          name: `bind`,
          arg: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: `style`,
            isStatic: true
          },
          exp: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: `_hoisted_1`,
            isStatic: false
          }
        },
        {
          type: NodeTypes.DIRECTIVE,
          name: `bind`,
          arg: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: `style`,
            isStatic: true
          },
          exp: {
            type: NodeTypes.COMPOUND_EXPRESSION,
            children: [
              {
                type: NodeTypes.SIMPLE_EXPRESSION,
                isStatic: false,
                content: 'true'
              },
              "?",
              createObjectExpression([
                createObjectProperty('display', createSimpleExpression('none', true))
              ]),
              ":",
              createObjectExpression([])
            ]
          }
        }
      ])
    })
  })
})
