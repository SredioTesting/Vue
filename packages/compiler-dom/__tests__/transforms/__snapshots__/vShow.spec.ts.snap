// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler: \`v-show\` transform should add style 1`] = `
Object {
  "children": Array [],
  "codegenNode": undefined,
  "isSelfClosing": true,
  "loc": Object {
    "end": Object {
      "column": 21,
      "line": 1,
      "offset": 20,
    },
    "source": "<div v-show=\\"true\\"/>",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "ns": 0,
  "props": Array [
    Object {
      "arg": undefined,
      "exp": Object {
        "content": "true",
        "isStatic": false,
        "loc": Object {
          "end": Object {
            "column": 18,
            "line": 1,
            "offset": 17,
          },
          "source": "true",
          "start": Object {
            "column": 14,
            "line": 1,
            "offset": 13,
          },
        },
        "type": 4,
      },
      "loc": Object {
        "end": Object {
          "column": 19,
          "line": 1,
          "offset": 18,
        },
        "source": "v-show=\\"true\\"",
        "start": Object {
          "column": 6,
          "line": 1,
          "offset": 5,
        },
      },
      "modifiers": Array [],
      "name": "show",
      "type": 7,
    },
  ],
  "tag": "div",
  "tagType": 0,
  "type": 1,
}
`;

exports[`compiler: \`v-show\` transform should append to style 1`] = `
Object {
  "children": Array [],
  "codegenNode": undefined,
  "isSelfClosing": true,
  "loc": Object {
    "end": Object {
      "column": 42,
      "line": 1,
      "offset": 41,
    },
    "source": "<div style=\\"display:flex\\" v-show=\\"true\\"/>",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "ns": 0,
  "props": Array [
    Object {
      "arg": Object {
        "content": "style",
        "isStatic": true,
        "loc": Object {
          "end": Object {
            "column": 26,
            "line": 1,
            "offset": 25,
          },
          "source": "style=\\"display:flex\\"",
          "start": Object {
            "column": 6,
            "line": 1,
            "offset": 5,
          },
        },
        "type": 4,
      },
      "exp": Object {
        "content": "_hoisted_1",
        "isStatic": false,
        "loc": Object {
          "end": Object {
            "column": 26,
            "line": 1,
            "offset": 25,
          },
          "source": "style=\\"display:flex\\"",
          "start": Object {
            "column": 6,
            "line": 1,
            "offset": 5,
          },
        },
        "type": 4,
      },
      "loc": Object {
        "end": Object {
          "column": 26,
          "line": 1,
          "offset": 25,
        },
        "source": "style=\\"display:flex\\"",
        "start": Object {
          "column": 6,
          "line": 1,
          "offset": 5,
        },
      },
      "modifiers": Array [],
      "name": "bind",
      "type": 7,
    },
    Object {
      "arg": undefined,
      "exp": Object {
        "content": "true",
        "isStatic": false,
        "loc": Object {
          "end": Object {
            "column": 39,
            "line": 1,
            "offset": 38,
          },
          "source": "true",
          "start": Object {
            "column": 35,
            "line": 1,
            "offset": 34,
          },
        },
        "type": 4,
      },
      "loc": Object {
        "end": Object {
          "column": 40,
          "line": 1,
          "offset": 39,
        },
        "source": "v-show=\\"true\\"",
        "start": Object {
          "column": 27,
          "line": 1,
          "offset": 26,
        },
      },
      "modifiers": Array [],
      "name": "show",
      "type": 7,
    },
  ],
  "tag": "div",
  "tagType": 0,
  "type": 1,
}
`;
